File: 2021/lib/day01.rb
# https://adventofcode.com/2021/day/1

class Day01
  def self.count_increasing(input, window_size)
    j = window_size
    count = 0
    input.each_with_index do |n, i|
      break if j == input.size
      m = input[j]
      count += 1 if m > n
      j += 1
    end
    count
  end

  def self.part1(input)
    count_increasing(input, 1)
  end

  def self.part2(input)
    count_increasing(input, 3)
  end
end

File: 2021/lib/day01.exs
# Answer based on Jose Valim's stream
# I'm learning elixir

{_, input} = File.read("input/day01.txt")

part1 = input
|> String.splitter("\n", trim: true)
|> Stream.map(&String.to_integer/1)
|> Stream.chunk_every(2, 1, :discard)
|> Enum.count(fn [left, right] -> right > left end)

part2 = input
|> String.splitter("\n", trim: true)
|> Stream.map(&String.to_integer/1)
|> Stream.chunk_every(4, 1, :discard)
|> Enum.count(fn [left, _, _, right] -> right > left end)

IO.write("Part 1: ")
IO.puts(part1)
IO.write("Part 2: ")
IO.puts(part2)
